plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -1574,16 +1576,67 @@ public void setColumnType(ConnectorSession session, ConnectorTableHandle tableHa
         verify(column.isBaseColumn(), "Cannot change nested field types");
 
         Table icebergTable = catalog.loadTable(session, table.getSchemaTableName());
+        Type sourceType = icebergTable.schema().findType(column.getName());
+        Type newType = toIcebergType(type);
         try {
-            icebergTable.updateSchema()
-                    .updateColumn(column.getName(), toIcebergType(type).asPrimitiveType())
+            buildUpdateSchema(column.getName(), sourceType, newType, icebergTable.updateSchema())
                     .commit();
         }
         catch (RuntimeException e) {
             throw new TrinoException(ICEBERG_COMMIT_ERROR, "Failed to set column type: " + firstNonNull(e.getMessage(), e), e);
         }
     }
 
+    private static UpdateSchema buildUpdateSchema(String name, Type sourceType, Type newType, UpdateSchema schema)
+    {
+        if (newType.isPrimitiveType()) {
+            return schema.updateColumn(name, newType.asPrimitiveType());
+        }
+        if (sourceType instanceof StructType sourceRowType && newType instanceof StructType newRowType) {
+            // Add, update or delete fields
+            for (NestedField field : concat(sourceRowType.fields(), newRowType.fields())) {
```

Comment:
This will have duplicates for fields in both old and new type. Also doesn't matter functionality wise but would be nice not to have to iterate over them twice.